Cleanup

gcie opened the pull request.

gcie committed:
fix: segmentation bug, allow for longer sentencepieces

gcie committed:
feat: cosine kmeans

gcie committed:
feat: clustering

gcie committed:
cleanup, proper readme

gcie committed:
docs: readme update

gcie committed:
feat: pieces statistics in segmentation

zsh01 commented on clusterization.py:
```diff
@@ -1,21 +1,38 @@
+                        help='Path to the dataset, which is to be quantized')
+    parser.add_argument('output', type=pathlib.Path,
+                        help='Output path')
+    parser.add_argument('--file_extension', type=str, default='vaw',
```
```suggestion
    parser.add_argument('--file-ext', type=str, default='wav',
```

zsh01 commented on README.md:
```diff
@@ -1 +1,49 @@
+
+**Example command**: `./scoring/run.sh`
+
+We compute the PER (Phonetic Error Rate) of a clusterization by mapping every sentence piece to the phone (from ground-truth), that occured most often at it's time. Then we do a greedy alignment, and compute the mismatch error rate.
```
```suggestion
We compute the PER (Phone Error Rate) of a clusterization by greedy-mapping every sentence piece to a most frequent ground-truth phone. Then we do a greedy alignment, and compute the mismatch error rate.
```

zsh01 commented on README.md:
```diff
@@ -1 +1,49 @@
+- `quantized`: path to the segmentation you want to rate
+- `frame_shift`: offset in number of frames, in most cases it should be 0
+
+## Clusterization
```
```suggestion
## Clustering
```

zsh01 commented on README.md:
```diff
@@ -1 +1,49 @@
+
+**Example command**: `./simi/grouping/run.sh`
+
+The idea is based on the fact that segmentation using sentencepiece requires large vocab to work, but it results in multiple different sentence pieces mapping to the same phoneme. In order to reduce number of sentence pieces, we run vord2vec on them and then kmeans, grouping and merging multiple sentencepieces together. It never increases PER, but (greatly) reduces vocab size.
```
```suggestion
The idea is based on the fact that segmentation using sentencepiece requires large vocab to work, but it results in multiple different sentence pieces mapping to the same phoneme. In order to reduce number of sentence pieces, we run word2vec on them and then k-means, grouping and merging multiple sentencepieces together. It never increases PER, but (greatly) reduces vocab size.
```

zsh01 commented on README.md:
```diff
@@ -1 +1,49 @@
+
+This is a script for running the CPC_kmeans checkpoint, but without the latest argmax (we leave each frame as distribution on centroids). This is required for the viterbi segmentation.
+
+To get the description of parameters run: `python clusterization.py --help`.
```
```suggestion
To get the description of parameters run: `python clustering.py --help`.
```

zsh01 commented on README.md:
```diff
@@ -1 +1,49 @@
+
+## Clusterization
+
+**Example command**: `./simi/clusterization/run.sh`
```
```suggestion
**Example command**: `./simi/clustering/run.sh`
```
Also rename the file.

zsh01 commented on clusterization.py:
```diff
@@ -1,21 +1,38 @@
+def parseArgs():
+    parser = ArgumentParser()
+
+    parser.add_argument('clustering_checkpoint', type=pathlib.Path,
```
```suggestion
    parser.add_argument('clustered-cpc-checkpoint', type=pathlib.Path,
```
Did I get it right - is it really a CPC + k-means checkpoint?

gcie replied:
Yes, it is basically a copy-paste from [zerospeech2021_baseline/scripts/quantize_audio.py](https://github.com/bootphon/zerospeech2021_baseline/blob/master/scripts/quantize_audio.py) without last argmax for we need the distribution on centroids.

zsh01 commented on clusterization.py:
```diff
@@ -1,21 +1,38 @@
+                        help='Output path')
+    parser.add_argument('--file_extension', type=str, default='vaw',
+                        help='File extension of the audio files')
+    parser.add_argument('--CUDA', action='store_true', 
```
```suggestion
    parser.add_argument('--cuda', action='store_true', 
```

zsh01 commented on clusterization.py:
```diff
@@ -1,21 +1,38 @@
+
+def quantize_file(file_path, cpc_feature_function, clusterModule, args):
     # Get CPC features
     cFeatures = cpc_feature_function(file_path)
```
Lets stick with underscored_case (camelCase in Python is used for class names).

zsh01 commented on clusterization.py:
```diff
@@ -27,17 +44,17 @@ def quantize_file(file_path, cpc_feature_function, clusterModule):
 
     clustered = clusterModule(cFeatures)
-    if CUDA:
+    if args.CUDA:
```
```suggestion
    if args.cuda:
```

zsh01 commented on clusterization.py:
```diff
@@ -27,17 +44,17 @@ def quantize_file(file_path, cpc_feature_function, clusterModule):
-    pathDB = '/pio/data/zerospeech2021/LibriSpeech/test-clean'
-    pathOutputDir = '/pio/scratch/1/i290956/zs2021/clusterings/LibriSpeech/test-clean'
+def main(args):
+    pathClusteringCheckpoint = str(args.clustering_checkpoint) # '/pio/data/zerospeech2021/checkpoints/CPC-big-kmeans50/clustering_kmeans50/clustering_CPC_big_kmeans50.pt'
```
Maybe lets throw all those paths into a .yaml config file? That would be clean & easy to switch between users.
.yamls are fun; unlike in .jsons, you can put comments.

zsh01 commented on clusterization.py:
```diff
@@ -27,17 +44,17 @@ def quantize_file(file_path, cpc_feature_function, clusterModule):
 
-    file_extension = 'vaw'
+    file_extension = args.file_extension
     seqNames, _ = findAllSeqs(pathDB, speaker_level=1, extension=file_extension, loadCache=True)
```
```suggestion
    seqNames, _ = findAllSeqs(pathDB, speaker_level=1, extension=args.file_ext, loadCache=True)
```

zsh01 commented on clusterization.py:
```diff
@@ -27,17 +44,17 @@ def quantize_file(file_path, cpc_feature_function, clusterModule):
+    pathOutputDir = str(args.output) # '/pio/scratch/1/i290956/zs2021/clusterings/LibriSpeech/test-clean'
 
-    file_extension = 'vaw'
+    file_extension = args.file_extension
```
```suggestion
```

zsh01 commented on clusterization.py:
```diff
@@ -68,7 +85,7 @@ def main():
     print(f"Loading ClusterModule at {pathClusteringCheckpoint}")
     clusterModule = loadClusterModule(pathClusteringCheckpoint)
-    if CUDA:
+    if args.CUDA:
```
```suggestion
    if args.cuda:
```

zsh01 commented on clusterization.py:
```diff
@@ -88,7 +105,7 @@ def main():
     if not clustering_args.train_mode:
         featureMaker.eval()
-    if CUDA:
+    if args.CUDA:
```
```suggestion
    if args.cuda:
```
After the fourth manual fix I already regret changing the case.

zsh01 commented on grouping.py:
```diff
@@ -0,0 +1,100 @@
+                        help='Path to the word2vec model, if not specified/empty then it will be computed')
+    parser.add_argument('--kmeans_path', type=str,
+                        help='Path to the kmeans model, if not specified/empty then it will be computed')
+    parser.add_argument('--seed', type=int, default=290956,
```
;)

zsh01 commented on grouping.py:
```diff
@@ -0,0 +1,100 @@
+    return parser.parse_args()
+
+
+class LibriSpeechSegmentation(object):
```
```suggestion
class LibriSpeechSegmentation:
```

zsh01 commented on grouping.py:
```diff
@@ -0,0 +1,100 @@
+        for root, _, files in os.walk(path):
+            for file in files:
+                if file.endswith('.csv'):
+                    for line in open(os.path.join(root, file), 'r', encoding='utf8'):
```
We've got `pathlib` imported, lets stick to it (`Path(root, file)`).

zsh01 commented on grouping.py:
```diff
@@ -0,0 +1,100 @@
+    print('Done!')
+
+
+class StubArgs(object):
```
.yaml it?

gcie replied:
This should not have been commited, I will remove it.

zsh01 commented on results/clustering_results.md:
```diff
@@ -0,0 +1,22 @@
+dataset | vocab size (pre) | PER (pre) sp | PER (pre) viterbi | vocab size (after) | PER (after) sp | PER (after) viterbi
```
Keeping track of the results in .md table - nice!

zsh01 commented on simi/clusterization/__init__.py:
```diff
@@ -0,0 +1,2 @@
+from .clusterization import *
```
Nooo lets not `from X import *` unless absolutely necessary.

gcie replied:
I wanted to be able to import anything from this folder by:
```python
from simi.clusterization import some_func
```
And not by:
```python
from simi.clusterization.clusterization import some_func
```

Is there any other way to achieve this?

alancucki replied:
I'm name things properly, e.g., with `quantization` we don't need the subdir.
Before:
`simi/quantization/quantization.py`
After:
`simi/quantization.py`
etc.

zsh01 commented on simi/clusterization/clusterization.py:
```diff
@@ -13,13 +13,19 @@ def cluster_kmeans(data, weights, path, n_clusters, train=True):
             raise Exception(f"Tried to cluster data, but there is no kmeans model at {path}. Maybe set train=True?")
         # run k-means
         print("Running kmeans...", flush=True)
+        if cosine:
```
Factor out `cosine( )` to a function?

zsh01 commented on simi/grouping/run.sh:
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+: ${VS:="10000 20000 50000 "}
```
Can we rename `VS` and `VVS` to sth more self-evident?

gcie replied:
Maybe remove the `VS` and rename `VVS` to `VS` or `VOCAB_SIZE`? This should be an example script, so maybe parameterizing the segmentation path is a bit unnecessary...

Also, the same would apply to `ALPHA`.

alancucki replied:
+1 for `VOCAB_SIZE`

zsh01 commented on simi/segmentation/segmentation.py:
```diff
@@ -34,19 +31,20 @@ def rate_segmentation(gt_segs, es_segs, tolerance=2):
         self.phone_list = set()
-        for root, dirs, files in os.walk(os.path.join(ALIGNMENTS_ROOTPATH, name)):
+        path = os.path.join(ALIGNMENTS_ROOTPATH, name) if path is None else path
+        for root, dirs, files in os.walk(path):
```
iirc there's something like `rglob()` in `Pathlib`:
```python
for file in Path(path).rglob('*.csv'):
    # one-level indent, yay!
```

zsh01 replied:
Also, `file` is a reserved keyword in Python.

gcie replied:
This is very clean solution! I also noticed that there is a similar issue in `grouping.py` file.

zsh01 commented on simi/statistics/pieces.py:
```diff
@@ -0,0 +1,53 @@
+
+
+def print_statistics(d, nl='\t'):
+    print(f"{nl}Total different pieces: {len(d.keys())}", \
```
If you've got parentheses, you don't need `\` to continue in the next line:
```python
print(f"{nl}Total different pieces: {len(d.keys())} "
      f"Total pieces: \t\t{sum(d.values())}")
```

zsh01 reviewed (commented):
Nice, thanks!

One think I'd add is poor man's configuration of bash scripts. In Python you would make a config .yaml or .json with all paths. In bash you can do:
```bash

# config.sh
CPC_MODEL=/path/to/my/cpc/model
CLUSTERINGS=/a/very/long/cymes/path/to/clusterings

# run.sh
source config.sh
python run_something.py $CLUSTERINGS $CPC_MODEL
```
In this way ppl can easily switch `config.sh` or lines inside it.

gcie committed:
Apply suggestions from code review

Co-authored-by: zsh01 <85645417+zsh01@users.noreply.github.com>

gcie committed:
Apply suggestions from code review comments

gcie commented:
I have addressed most of the issues and ideas except for yaml-ing the `clusterization.py`. It is basically a copy-paste from  [zerospeech2021_baseline/scripts/quantize_audio.py](https://github.com/bootphon/zerospeech2021_baseline/blob/master/scripts/quantize_audio.py) and I don't think it will be run too often - it prepares dataset for you to work on.
